### PR TITLE
Modify root volume size without instance recreation

### LINKED ISSUE #768: Changing root_block_device.volume_size forces new resource (aws_instance)
_This issue was originally opened by @cweagans as hashicorp/terraform#14445. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

### Terraform Version
v0.9.5

### Affected Resource(s)
- aws_instance

#### Before:
```hcl
provider "aws" {
  region = "us-east-1"
}

data "aws_ami" "ubuntu" {
  most_recent = true

  filter {
    name = "name"
    values = ["ubuntu/images/hvm-ssd/ubuntu-trusty-14.04-amd64-server-*"]
  }

  filter {
    name = "virtualization-type"
    values = ["hvm"]
  }

  owners = ["099720109477"]
}

resource "aws_instance" "test_instance" {
  ami = "${data.aws_ami.ubuntu.id}"
  instance_type = "t2.micro"

  tags {
    Name = "cweagans-test-instance"
  }
}
```

#### After:
```hcl
provider "aws" {
  region = "us-east-1"
}

data "aws_ami" "ubuntu" {
  most_recent = true

  filter {
    name = "name"
    values = ["ubuntu/images/hvm-ssd/ubuntu-trusty-14.04-amd64-server-*"]
  }

  filter {
    name = "virtualization-type"
    values = ["hvm"]
  }

  owners = ["099720109477"]
}

resource "aws_instance" "test_instance" {
  ami = "${data.aws_ami.ubuntu.id}"
  instance_type = "t2.micro"

  tags {
    Name = "cweagans-test-instance"
  }

  root_block_device {
    volume_size = "100"
  }
}
```

### Expected Behavior
Root volume should have been resized from the default 8GB to 100GB without deleting the instance in the process.

### Actual Behavior
The instance was deleted and recreated.

### Steps to Reproduce
`terraform apply` the first config above, change config to the second one, and `terraform apply` again.

### LINKED ISSUE #1297: aws_volume_attachment force new resource when instance_id is computed
_This issue was originally opened by @notuscloud as hashicorp/terraform#15659. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

### Terraform Version

`v0.9.11`

### Terraform Configuration Files
```hcl
# EC2 instance
module "elk" {
  source = "../commons/aws/ec2/instance"
  ami = "${var.aws_default_ami}"
  type = "${var.elk_flavor}"
  key_name = "${var.aws_default_keyname}"
  subnet_id = "${data.aws_subnet.logs.id}"
  associate_public_ip_address = true
  availability_zone = "${var.availability_zone}"
  vpc_security_group_ids = ["${data.aws_security_group.logs_main_sg.id}"]
  docker_disk_size = 16
  tags {
    Name = "elk"
    env = "${var.env}"
    stack = "${var.stack}"
  }
}

# EBS Volume attachment
resource "aws_volume_attachment" "elk_ebs" {
  device_name  = "/dev/sdc"
  volume_id    = "${data.aws_ebs_volume.elk_data.id}"
  instance_id  = "${module.elk.id}"
  force_detach = true
}
```

### Debug Output

```
-/+ aws_volume_attachment.elk_ebs
    device_name:  "/dev/sdc" => "/dev/sdc"
    force_detach: "true" => "true"
    instance_id:  "i-someid" => "${module.elk.id}" (forces new resource)
    skip_destroy: "" => "<computed>"
    volume_id:    "vol-someid" => "vol-someid"

-/+ module.elk.aws_instance.instance
    ami:                                               "ami-fa2df395" => "ami-fa2df395"
    associate_public_ip_address:                       "true" => "true"
    availability_zone:                                 "eu-central-1a" => "eu-central-1a"
    ebs_block_device.#:                                "2" => "1"
    ebs_block_device.2554893574.delete_on_termination: "false" => "false"
    ebs_block_device.2554893574.device_name:           "/dev/sdc" => "" (forces new resource)
    ebs_block_device.2576023345.delete_on_termination: "true" => "true"
    ebs_block_device.2576023345.device_name:           "/dev/sdb" => "/dev/sdb"
    ebs_block_device.2576023345.encrypted:             "false" => "<computed>"
    ebs_block_device.2576023345.iops:                  "0" => "<computed>"
    ebs_block_device.2576023345.snapshot_id:           "" => "<computed>"
    ebs_block_device.2576023345.volume_size:           "16" => "16"
    ebs_block_device.2576023345.volume_type:    
```

### Expected Behavior

`aws_volume_attachment` should not be forced into a new resource, thus the drive letter should not change for the instance. As nothing changes for the attachment and aws_instance, nothing should happen.

### Actual Behavior

* Terraform detects a change with the instance_id of the `aws_volume_attachment` and force a new resource
* This new resource thing impacts the associated `aws_instance`
I suspect that the `aws_volume_attachment` being forced anew, it provokes a device_name change to the `aws_instance`.
* This change to the `device_name` forces a new resource with the attached `aws_instance`

To resume, I do no changes and both `aws_volume_attachment` and `aws_instance` are being deleted then re-created.

### Steps to Reproduce

1. `terraform apply`
2. Do nothing
3. `terraform plan` to see what's gonna happen or `terraform apply`

Thanks for your help,
Thomas

### PR DESCRIPTION
**Pull request by viscat**
This is my first PR to Terraform, and I guess it will not be the last one. Terraform (also Packer) helped me a lot, and I think it's time to contribute a little :)

Using the recently AWS feature to modify root volume size without forcing new instance, I added a call to AWS API to modify it on the fly. I added an acceptance test to check the volume size update is done without instance recreation.

Fixes #768

### CODE REVIEW

**Comment by ricardclau** [comment]
Hi @radeksimko any news here? 

We have tested this with our AWS setup and it seems to work fine, not sure if there is anything missing in the PR for you guys to review and merge

**Comment by mbelang** [comment]
Any news on this? We have some needs for that feature.

File: aws/resource_aws_instance.go [hunk 1]
```diff
@@ -1011,6 +1010,58 @@ func resourceAwsInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
...
+	if d.HasChange("root_block_device") {
+		if rbd, ok := d.GetOk("root_block_device"); ok {
+			rbdL := rbd.([]interface{})
+			if len(rbdL) > 1 {
+				return fmt.Errorf("Cannot specify more than one root_block_device.")
+			}
+			for _, r := range rbdL {
+				bd := r.(map[string]interface{})
+				if volumeSize, ok := bd["volume_size"].(int); ok && volumeSize != 0 {
+					volumeIds, err := getAwsInstanceVolumeIds(conn, d)
```
Comment (radeksimko) [inline aws/resource_aws_instance.go]: Do we know for sure that the first volume in the list is always RBD? The docs don't make it very clear 🤔  http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeVolumes.html and we'd rather not resize wrong volume by accident.

File: aws/resource_aws_instance.go [hunk 2]
```diff
@@ -1011,6 +1010,58 @@ func resourceAwsInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
...
 
+	if d.HasChange("root_block_device") {
+		if rbd, ok := d.GetOk("root_block_device"); ok {
+			rbdL := rbd.([]interface{})
+			if len(rbdL) > 1 {
+				return fmt.Errorf("Cannot specify more than one root_block_device.")
+			}
+			for _, r := range rbdL {
+				bd := r.(map[string]interface{})
+				if volumeSize, ok := bd["volume_size"].(int); ok && volumeSize != 0 {
```
Comment (radeksimko) [inline aws/resource_aws_instance.go]: I think you can simplify a lot of the code above by using `d.HasChange("root_block_device.0.volume_size")` and `d.Get("root_block_device.0.volume_size")` - you can omit the loop, validation, some of the casting etc.

File: aws/resource_aws_instance.go [hunk 3]
```diff
@@ -1066,6 +1117,35 @@ func InstanceStateRefreshFunc(conn *ec2.EC2, instanceID, failState string) resou
...
 
+// VolumeStateRefreshFunc returns a resource.StateRefreshFunc that is used to watch
+// an EC2 root device volume.
+func VolumeStateRefreshFunc(conn *ec2.EC2, volumeID, failState string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		resp, err := conn.DescribeVolumesModifications(&ec2.DescribeVolumesModificationsInput{
+			VolumeIds: []*string{aws.String(volumeID)},
+		})
+		if err != nil {
+			if ec2err, ok := err.(awserr.Error); ok && ec2err.Code() == "InvalidVolumeID.NotFound" {
```
Comment (radeksimko) [inline aws/resource_aws_instance.go]: We have a nice helper for these kind of checks - see https://github.com/terraform-providers/terraform-provider-aws/blob/b8d4e1570fc43a2acee6b6e47f63c9db6b067fa2/aws/resource_aws_emr_security_configuration.go#L101

File: aws/resource_aws_instance.go [hunk 4]
```diff
@@ -1066,6 +1117,35 @@ func InstanceStateRefreshFunc(conn *ec2.EC2, instanceID, failState string) resou
...
+				resp = nil
```
Comment (radeksimko) [inline aws/resource_aws_instance.go]: Why can't we just `return nil, "", nil` here? That would also simplify the code below as we can then omit the else block and conditional below. 😉

File: aws/resource_aws_instance_test.go [hunk 5]
```diff
@@ -1016,6 +1016,35 @@ func TestAccAWSInstance_changeInstanceType(t *testing.T) {
 	})
 }
 
+func TestAccAWSInstance_changeRootBlockDeviceVolumeSize(t *testing.T) {
```
Comment (radeksimko) [inline aws/resource_aws_instance_test.go]: This test is currently failing:

```
=== RUN   TestAccAWSInstance_changeRootBlockDeviceVolumeSize
--- FAIL: TestAccAWSInstance_changeRootBlockDeviceVolumeSize (311.43s)
	testing.go:434: Step 1 error: Check failed: Check 3/3 error: aws_instance.foo: Attribute 'root_block_device.0.volume_size' expected "20", got "12"
FAIL
exit status 1
```

File: aws/resource_aws_instance.go [hunk 6]
```diff
@@ -1011,6 +1010,58 @@ func resourceAwsInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
...
+
+					volResp, err := conn.DescribeVolumes(&ec2.DescribeVolumesInput{
+						VolumeIds: volumeIds,
+					})
+					if err != nil {
+						return err
+					}
+
+					if len(volResp.Volumes) < 1 {
+						return fmt.Errorf("Cannot fetch volume info")
```
Comment (radeksimko) [inline aws/resource_aws_instance.go]: Do you mind making the error message a bit more helpful here? e.g. `"No volumes found for %s", d.Id()`

**Comment by viscat** [comment]
Thank you very much @radeksimko for all the feedback. I will also check the conflicts with master branch. I  will fix and improve what you pointed out as soon as possible.

**Comment by johntdyer** [comment]
Any update on merging this ?

**Comment by radeksimko** [comment]
👋  @viscat 
do you think you'll have time to look at this PR or are you happy for anyone else to pick it up?

Thanks.

**Comment by bflad** [comment]
Hi @viscat 👋  will you have time to implement the feedback above? If not, just let us know. Thanks.

**Comment by siliconavengers** [comment]
I hope this PR will be merged soon

**Comment by viscat** [comment]
Sorry @bflad right now I can't do it.

**Comment by viscat** [comment]
@radeksimko @bflad  Sorry for taking so long to take it again. I made the fixes and improved that you requested. Thank you for the feedback! The only thing I am not able to figure out is how to determine which is the RBD. I have been looking the AWS API doc and I can't find anything. Can you help me with this please? I can try to get in touch with AWS support anyways.

**Comment by viscat** [comment]
@radeksimko Can we merge this?

**Comment by dimisjim** [comment]
@bflad Hey guys, what's stopping us from merging this?

**Comment by gdavison** [comment]
Hi @viscat, thanks for your patience. I completely understand if you're not wanting to continue working on this, and we can take it over if you want.

If you are still interested, I have two requests:
* It looks like some of our internal function have changed their interfaces, and this PR no longer builds against the current provider code. Can you rebase it, please?
* If there are multiple EBS volumes attached to an instance, how can we guarantee that we're getting the correct volume for resizing? We might be able to use the volume id in the state.

**Comment by gdavison** [comment]
I've created #12620 to continue this work

**Pull request closed without merging**